OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,302 @@
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStopping.
+  CaptureStatus prev_capture_status = CaptureStatus::kCaptureFinished;
+
+  while (true) {
+    // This is set when ReceiveEventsThread has exited. Then, this thread should also exit.
+    if (*exit_send_commands_thread) {
+      return;
+    }
+
+    service_state_mutex_.Lock();
+    if (service_state_.exit_requested) {
+      service_state_mutex_.Unlock();
+      return;
+    }
+
+    if (service_state_.capture_status != prev_capture_status) {
+      switch (service_state_.capture_status) {
+        case CaptureStatus::kCaptureStarted: {
+          prev_capture_status = service_state_.capture_status;
+          ++service_state_.producers_remaining;
+          *all_events_sent_received = false;
+          service_state_mutex_.Unlock();
+
+          orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse command;
+          command.mutable_start_capture_command();
+          if (!stream->Write(command)) {
+            ERROR("Sending StartCaptureCommand to CaptureEventProducer");
+            LOG("Terminating call to ReceiveCommandsAndSendEvents as Write failed");
+            // Cause Read in ReceiveEventsThread to also fail if for some reason it hasn't already.
+            context->TryCancel();
+            return;
+          }
+          LOG("Sent StartCaptureCommand to CaptureEventProducer");
+        } break;
+
+        case CaptureStatus::kCaptureStopping: {
+          prev_capture_status = service_state_.capture_status;
+          service_state_mutex_.Unlock();
+
+          orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse command;
+          command.mutable_stop_capture_command();
+          if (!stream->Write(command)) {
+            ERROR("Sending StopCaptureCommand to CaptureEventProducer");
+            LOG("Terminating call to ReceiveCommandsAndSendEvents as Write failed");
+            // Cause Read in ReceiveEventsThread to also fail if for some reason it hasn't already.
+            context->TryCancel();
+            return;
+          }
+          LOG("Sent StopCaptureCommand to CaptureEventProducer");
+        } break;
+
+        case CaptureStatus::kCaptureFinished: {
+          prev_capture_status = service_state_.capture_status;
+          *all_events_sent_received = true;
+          service_state_mutex_.Unlock();
+        } break;
+      }
+      continue;
+    }
+
+    // Wait for service_state_.capture_status to change or for service_state->exit_requested
+    // (the next iteration will handle the change).
+    // Use a timeout to periodically check (in the next iteration)
+    // for *terminate_send_commands_thread, set by ReceiveCommandsAndSendEvents.
+    static constexpr absl::Duration kCheckExitSendCommandsThreadInterval = absl::Seconds(1);
+    switch (service_state_.capture_status) {
+      case CaptureStatus::kCaptureStarted: {
+        service_state_mutex_.AwaitWithTimeout(absl::Condition(
```

Comment:
It seems that you need this super duplicated code (for all the three cases), as you need a function pointer in the AwaitWithTimout, which can't capture the state at the beginning. If so, could you add comment for that?